### PR TITLE
conditional operators

### PR DESCRIPTION
**Pull request by hnyu**

### CODE REVIEW

File: alf/utils/conditional_ops.py [hunk 1]
```diff
@@ -62,58 +61,29 @@ def conditional_update(target, cond, func, *args, **kwargs):
     Returns:
         nest with the same structure and shape as target.
     """
-    # shape of indices from where() is [batch_size,1], which is what scatter_nd
-    # needs
-    scatter_indices = tf.where(cond)
-    scope = get_current_scope()
+    # shape of indices from where() is a tuple (indices, )
```
Comment (emailweixu) [inline alf/utils/conditional_ops.py]: wrong comment

File: alf/utils/conditional_ops.py [hunk 2]
```diff
@@ -62,58 +61,29 @@ def conditional_update(target, cond, func, *args, **kwargs):
...
-        # func unscoped. We need this line to restore the original name scope.
-        with tf.name_scope(scope):
-            func()
-            return tf.constant(True)
-
-    tf.cond(cond, _if_true, lambda: tf.constant(False))
+        updates = func(*selected_args, **selected_kwargs)
+
+        def _update(tgt, updt):
+            scatter_indices = torch.reshape(
```
Comment (emailweixu) [inline alf/utils/conditional_ops.py]: scatter_indices = common.expand_dims_as(gather_indices, updt)

File: alf/utils/conditional_ops.py [hunk 3]
```diff
@@ -62,58 +61,29 @@ def conditional_update(target, cond, func, *args, **kwargs):
...
+                gather_indices,
+                list(gather_indices.shape) +
+                [-1] * (updt.ndim - gather_indices.ndim))
+            scatter_indices = scatter_indices.expand_as(updt)
+            return tgt.scatter(0, scatter_indices, updt)
```
Comment (emailweixu) [inline alf/utils/conditional_ops.py:76]: I wonder whether this more readable way work (whether gradient is correctly handled):
```python
ret = tgt.clone()
ret[gather_indices] = updt
return ret
```

Comment (hnyu) [reply alf/utils/conditional_ops.py:76]: For some pytorch operators, in-place operations will throw errors.

File: alf/utils/conditional_ops.py [hunk 4]
```diff
@@ -62,58 +62,26 @@ def conditional_update(target, cond, func, *args, **kwargs):
     Returns:
         nest with the same structure and shape as target.
     """
-    # shape of indices from where() is [batch_size,1], which is what scatter_nd
-    # needs
-    scatter_indices = tf.where(cond)
-    scope = get_current_scope()
+    # shape of torch.where() is a tuple (indices, )
```
Comment (emailweixu) [inline alf/utils/conditional_ops.py]: shape is not a tuple.
The return value of torch.where() is a tuple (indices, )

Comment (hnyu) [reply alf/utils/conditional_ops.py]: I guess here 'shape' means in the general sense (what it looks like). I can update it.

**Review by emailweixu — APPROVED** [review]
